### PR TITLE
Add how-to for boundary conditions

### PR DESCRIPTION
**Pull request by charleskawczynski**
### Description

This PR adds a how-to for defining boundary conditions in ClimateMachine.

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+## Discontinuous Galerkin (DG) method
+
+Boundary conditions (BCs) are defined _weakly_ in ClimateMachine's
+DG implementation. We do this by defining BCs _through_ the _fluxes_,
+rather than on the prognostic variables themselves. Therefore, BCs
+are enforced by prescribing fluxes (numerical fluxes, in DG terminology)
+that drive the solution at the boundary to satisfy our desired BCs.
+
+One major advantage of this implementation approach is that BCs are
+handled in phase space, so we only need one implementation for all
```
Comment (LenkaNovak) [inline docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md:16]: may be useful to clarify what is meant by phase space here. 🤔

File: docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+[`BalanceLaw`](@ref) to fully define BCs:
+
+```@meta
+CurrentModule = ClimateMachine.DGMethods.NumericalFluxes
+```
+
+| **Method**                                       | Internal/User-facing  |  If present | Calls to user-facing functions |
+|:-----|:-----|:-----|:-----|
+| [`numerical_boundary_flux_first_order!`](@ref)   | User-facing  | F₁                  | `boundary_state!(::NumericalFluxFirstOrder)`, `numerical_flux_first_order!(::NumericalFluxFirstOrder)`    |
+| [`numerical_boundary_flux_second_order!`](@ref)  | User-facing  | F₂                  | `normal_boundary_flux_second_order!`[, `boundary_flux_second_order!` ,`boundary_state!(::NumericalFluxSecondOrder)`, `flux_second_order!`]  |
```
Comment (LenkaNovak) [inline docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md:33]: I believe in the code the (normal) `local_flux` = `fluxTn` = F1+F2 = the total flux, so when you're modifying it in the `numerical_boundary_flux_second_order` you need to be careful not to overwrite the first flux contribution (unless that's what you want to do, i.e. impose the Robin BC on the total flux)

File: docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# How to define boundary conditions
```
Comment (LenkaNovak) [inline docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md:1]: Should we also include the BC specification using the Dirichlet /  Neumann BCs in `boundary_state!` as is done in the Land Tutorial?

**Review by LenkaNovak — COMMENTED** [review]
Looks good. Just a few comments:

File: docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+`R(Y, t) = ∂_t Yᵢ + (∇•(F₁(Y) + F₂(Y,G)))ᵢ - (S(Y,G))ᵢ = 0, G = ∇Y`,
+
+the following functions must be defined for a
+[`BalanceLaw`](@ref) to fully define BCs:
+
+```@meta
+CurrentModule = ClimateMachine.DGMethods.NumericalFluxes
+```
+
+| **Method**                                       | Internal/User-facing  |  If present | Calls to user-facing functions |
```
Comment (jb-mackay) [inline docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md:30]: It's not immediately clear to me what is meant by "Calls to user-facing functions" and "Internal/User-facing" columns. They may need some clarification/motivation outside the table -- why does the user need this information and how will they use it when defining BCs?

File: docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+One major advantage of this implementation approach is that BCs are
+handled in phase space, so we only need one implementation for all
+time steppers.
+
+In the context of the PDE that `ClimateMachine.jl` solves:
+
+`R(Y, t) = ∂_t Yᵢ + (∇•(F₁(Y) + F₂(Y,G)))ᵢ - (S(Y,G))ᵢ = 0, G = ∇Y`,
+
+the following functions must be defined for a
```
Comment (jb-mackay) [inline docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md:23]: "_must_ be defined" feels at odds with the "If present" column in the table. Perhaps something like "The terms in a balance law determine what functions must be defined for the BCs..." and making the "If present" column the second column?

File: docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+∫_D R(Y, t) l(x) dx = ∫_∂D n̂ • (f^{-} - f^{*}) l(x) dx
+``
+
+Where ``f^{-}`` is the surface value on the ``-`` state (the interior
+element), and ``f^{*}`` is the prescribed numerical flux that satisfies
+our BCs.
+
+This requires us to translate our boundary conditions.
+It is often desired to use the same numerical flux stencil on the interior
+ as the boundary.
```
Comment (jb-mackay) [inline docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md:51]: Does this belong after the table or towards the start when discussing the weak implementation of BCs? I think the line about translating boundary conditions needs a little more context.

File: docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md (hunk 2 above)
Comment (charleskawczynski) [reply docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md:33]: Maybe this would be a good footnote somewhere?

File: docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md (hunk 4 above)
Comment (charleskawczynski) [reply docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md:30]: We can add more information. The basic idea is that the methods in the far left column, by default, call the methods to the right, which, by default, call methods to the right of them. This means that users can intercept these calls at any point along the call graph, depending on how they want to specify the BCs-- for example by specifying the numerical flux directly, via the far left column method or only the ghost points (via `boundary_state!`).

File: docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md (hunk 5 above)
Comment (charleskawczynski) [reply docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md:23]: Agreed, this needs more work.

File: docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md (hunk 6 above)
Comment (charleskawczynski) [reply docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md:51]: Agreed, this needs more thought.

File: docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md (hunk 3 above)
Comment (charleskawczynski) [reply docs/src/HowToGuides/BalanceLaws/how_to_define_bcs.md:1]: Well, I was hoping to cover BC implementation in detail in the BC tutorial

**Comment by jkozdon** [comment]
@charleskawczynski  can you close this? Seems pretty stale.
